sw/banshee/src/engine.rs

```diff
@@ -647,23 +703,74 @@ impl<'a, 'b> Cpu<'a, 'b> {
         prev as u32
     }
 
-    fn binary_csr_read(&self, csr: u16) -> u32 {
-        trace!("Read CSR 0x{:x}", csr);
+    fn binary_csr_read(&self, csr: riscv::Csr) -> u32 {
+        trace!("Read CSR {:?}", csr);
         match csr {
-            0x7C0 => self.state.ssr_enable,
-            0xB00 => self.state.cycle as u32,         // csr_mcycle
-            0xB80 => (self.state.cycle >> 32) as u32, // csr_mcycleh
-            0xB02 => self.state.instret as u32,       // csr_minstret
-            0xB82 => (self.state.instret >> 32) as u32, // csr_minstreth
-            0xF14 => self.hartid as u32,              // mhartid
+            riscv::Csr::Ssr => self.state.ssr_enable,
+            riscv::Csr::Mcycle => self.state.cycle as u32, // csr_mcycle
+            riscv::Csr::Mcycleh => (self.state.cycle >> 32) as u32, // csr_mcycleh
+            riscv::Csr::Minstret => self.state.instret as u32, // csr_minstret
+            riscv::Csr::Minstreth => (self.state.instret >> 32) as u32, // csr_minstreth
+            riscv::Csr::Mhartid => self.hartid as u32,     // mhartid
+            riscv::Csr::Mstatus => self.state.irq.mstatus, // CSR_MSTATUS
+            riscv::Csr::Mie => self.state.irq.mie,         // CSR_MIE
+            riscv::Csr::Mip => self.state.irq.mip,         // CSR_MIP
+            riscv::Csr::Mtvec => self.state.irq.mtvec,     // CSR_MTVEC
+            riscv::Csr::Mepc => self.state.irq.mepc,       // CSR_MEPC
+            riscv::Csr::Mcause => self.state.irq.mcause,   // CSR_MCAUSE
+            riscv::Csr::Misa => {
+                // RV32IMAFDX A - Atomic Instructions extension
+                (1 << 0) | (1 << 3) | (1 << 5) | (1 << 8) | (1 << 12) | (1 << 23) | (1 << 30)
+            }
             _ => 0,
         }
     }
 
-    fn binary_csr_write(&mut self, csr: u16, value: u32) {
-        trace!("Write CSR 0x{:x} = 0x{:?}", csr, value);
+    fn binary_csr_write(&mut self, csr: riscv::Csr, value: u32) {
+        trace!("Write CSR {:?} = 0x{:?}", csr, value);
         match csr {
-            0x7C0 => self.state.ssr_enable = value,
+            riscv::Csr::Ssr => self.state.ssr_enable = value,
+            riscv::Csr::Mstatus => self.state.irq.mstatus = value, // CSR_MSTATUS
+            riscv::Csr::Mie => self.state.irq.mie = value,         // CSR_MIE
+            riscv::Csr::Mip => self.state.irq.mip = value,         // CSR_MIP
+            riscv::Csr::Mtvec => self.state.irq.mtvec = value,     // CSR_MTVEC
+            riscv::Csr::Mepc => self.state.irq.mepc = value,       // CSR_MEPC
+            riscv::Csr::Mcause => self.state.irq.mcause = value,   // CSR_MCAUSE
+            _ => (),
+        }
+    }
+
+    fn binary_csr_read_silent(&self, csr: riscv::Csr) -> u32 {
```

Comment:
Is the only difference to the 'regular' car functions the tracing?

Reply:
yes the `trace!` is the only difference. Since with the interrupts the CSR are read on every instruction this would else pollute the output when running with `SNITCH_LOG=trace` (has nothing todo with the `--trace` tracer).

The redundant code is ugly, should I solve this with a function argument?